sgt-back-end/server.js

```diff
@@ -0,0 +1,172 @@
+import express from 'express';
+import pg from 'pg';
+
+// only create ONE pool for your whole server
+const db = new pg.Pool({
+  connectionString: 'postgres://dev:dev@localhost/studentGradeTable',
+  ssl: {
+    // Allow non-SSL traffic to localhost
+    rejectUnauthorized: false,
+  },
+});
+
+const app = express();
+
+// middleware to parse JSON request bodies
+app.use(express.json());
+
+// Defining a route to handle GET requests to retrieve all grades
+app.get('/api/grades', async (res) => {
+  try {
+    const sql = `
+      SELECT * FROM "grades"
+    `;
+    const result = await db.query(sql);
+
+    // Checking if any rows were returned from the query
+    const grades = result.rows;
+
+    if (grades.length === 0) {
+      // If there are no rows, return an empty array
+      res.status(200).json([]);
+    } else {
+      // If there are rows, return them as an array of objects
+      res.status(200).json(grades);
+    }
+  } catch (err) {
+    // Handling errors by sending a 500 status code and an error message
+    res.status(500).json({ error: 'An error occurred while fetching grades.' });
+  }
+});
+
+// Defining a route to handle POST requests to insert a new grade
+app.post('/api/grades', async (req, res) => {
+  try {
+    // Extracting data from the request body
+    const { name, course, score } = req.body;
+
+    // Validating the incoming data
+    if (!name || !course || isNaN(score) || score < 0 || score > 100) {
+      // Handling client validation errors with a 400 status code
+      res.status(400).json({ error: 'Invalid grade data.' });
+    } else {
+      // Inserting the new grade into the database
+      const sql = `
+        INSERT INTO "grades" ("name", "course", "score")
+        VALUES ($1, $2, $3)
+        ORDER BY "gradeId"
+        RETURNING *
+      `;
+      const params = [name, course, score];
+      const result = await db.query(sql, params);
+
+      // Retrieving the created grade
+      const createdGrade = result.rows[0];
+
+      // Responding with a 201 status code and the created grade
+      res.status(201).json(createdGrade);
+    }
+  } catch (err) {
+    // Handling server errors with a 500 status code
+    console.error('Error:', err); // Log the error for debugging
+    res
+      .status(500)
+      .json({ error: 'An error occurred while creating the grade.' });
+  }
+});
+
+// Defining a route to handle PUT requests to update a row by gradeId
+app.put('/api/grades/:gradeId', async (req, res) => {
+  try {
+    // Extracting data from the request body
+    const { name, course, score } = req.body;
+    const { gradeId } = req.params; // Extract gradeId from URL parameters
+
+    // Validating the incoming data
+    if (!name || !course || isNaN(score) || score < 0 || score > 100) {
+      // Handling client validation errors with a 400 status code
+      res.status(400).json({ error: 'Invalid grade data.' });
+    } else {
+      const checkGradeId = `SELECT * FROM "grades" 
```

Comment:
It's actually not necessary to do this select first. You can just issue the `update` query with a `where` clause and `returning *`. Think about what happens:
The database tries to update the row specified by the `where` clause. It doesn't find one, so it doesn't update anything. It therefore returns an empty array.

So you can instead issue the update query, check the result and if it's empty return a 404. If it's not empty, return a 200 and the result.